code/game/objects/items/RCD.dm

```diff
@@ -1157,56 +1182,152 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
-				new blueprint (A, FALSE, layers[current_layer])
+				playsound(loc, 'sound/machines/click.ogg', 50, TRUE)
+				if(ispath(blueprint, /obj/machinery/duct))
+					var/is_omni = current_color == DUCT_COLOR_OMNI
+					new blueprint(destination, FALSE, GLOB.pipe_paint_colors[current_color], GLOB.plumbing_layers[current_layer], null, is_omni)
+				else
+					new blueprint(destination, FALSE, GLOB.plumbing_layers[current_layer])
 				return TRUE
 
-/obj/item/construction/plumbing/proc/canPlace(turf/T)
-	if(!isopenturf(T))
+/obj/item/construction/plumbing/proc/canPlace(turf/destination)
+	if(!isopenturf(destination))
 		return FALSE
 	. = TRUE
-	for(var/obj/O in T.contents)
-		if(O.density) //let's not built ontop of dense stuff, like big machines and other obstacles, it kills my immershion
+
+	var/obj/blueprint_template = blueprint
+	var/layer_id = GLOB.plumbing_layers[current_layer]
+
+	for(var/obj/machinery/machine in destination.contents)
+		// Let's not built ontop of dense stuff, if this is also dense.
+		if(initial(blueprint_template.density) && machine.density)
 			return FALSE
 
-/obj/item/construction/plumbing/afterattack(atom/A, mob/user, proximity)
+		// Ducts and overlap machines IF the layers are different
+
+		// make sure plumbling isn't overlapping.
+		for(var/datum/component/plumbing/plumber as anything in machine.GetComponents(/datum/component/plumbing))
+			if(plumber.ducting_layer & layer_id)
+				return FALSE
+
+		if(istype(machine, /obj/machinery/duct))
+			// Make sure ducts aren't overlapping.
+			var/obj/machinery/duct/duct_machine = machine
+			if(duct_machine.duct_layer & layer_id)
+				return FALSE
+
+/obj/item/construction/plumbing/afterattack(atom/target, mob/user, proximity)
 	. = ..()
 	if(!prox_check(proximity))
 		return
-	if(istype(A, /obj/machinery/plumbing))
-		var/obj/machinery/plumbing/P = A
-		if(P.anchored)
-			to_chat(user, span_warning("The [P.name] needs to be unanchored!"))
+	if(istype(target, /obj/machinery/plumbing))
+		var/obj/machinery/machine_target = target
+		if(machine_target.anchored)
+			to_chat(user, span_warning("The [target.name] needs to be unanchored!"))
 			return
-		if(do_after(user, 20, target = P))
-			P.deconstruct() //Let's not substract matter
+		if(do_after(user, 20, target = target))
+			machine_target.deconstruct() //Let's not substract matter
 			playsound(get_turf(src), 'sound/machines/click.ogg', 50, TRUE) //this is just such a great sound effect
 	else
-		create_machine(A, user)
+		create_machine(target, user)
 
 /obj/item/construction/plumbing/AltClick(mob/user)
-	if(!istype(user) || !user.canUseTopic(src, BE_CLOSE))
+	// give a menu to pick layers or colors
+	var/list/options_menu = list(
+		"Current Layer" = image('icons/hud/radial.dmi', icon_state = "plumbing_layer[GLOB.plumbing_layers[current_layer]]"),
+		"Current Color" = image('icons/hud/radial.dmi', icon_state = current_color),
+	)
+
+	playsound(src.loc, 'sound/effects/pop.ogg', 50, FALSE)
```

Comment:
src.loc

Reply:
Replaced *all* instanced of `src.loc` in rcd.dm to loc since these playsound lines are just copy-pastes in the first place.